File: lib/scoring/text_entry.rb
# frozen_string_literal: true

module Scoring
  class TextEntry
    def calculate(_question, result, question_scoring)
      values = []
      scoring_template = question_scoring.props

      result['answers'].each do |answer|
        if answer['value']
          object = scoring_template.find { |template| template['index'] == answer['value'] }
          values << object['value'] if object && object['value']
        end
      end
      value = values.empty? ? nil : values.sum.to_f / values.size
      { value: value, options: [] }
    end
  end
end

File: app/frontend/modules/survey/models/Validations/Selectors/TextEntry.ts
import _ from 'lodash'

interface Answer {
  value: boolean | string | string[]
  index: number
}
interface Condition {
  answer: {}
  result: { answers: Answer[]}
  value: string
  type: string
}

function find (condition: Condition) {
  const answers = _.get(condition, ['result', 'answers'], [])
  const answer = _.find(answers, { index: +condition.answer })
  if (answer) {
    if (_.isArray(answer.value)) {
      if (answer.value.includes(condition.value)) {
        return answer
      }
    }
    if (answer.value === (condition.type === 'bool' ? true : condition.value)) {
      return answer
    }
  }
  return null
}

export default {
  Form (condition: Condition) {
    return find(condition)
  },
}
